kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SchemaManager.java

```diff
@@ -0,0 +1,305 @@
+          implements Comparator<FileStatus> {
+
+    @Override
+    public int compare(FileStatus file1, FileStatus file2) {
+
+      return getFileNumber(file1) - getFileNumber(file2);
+    }
+  }
+
+  /**
+   * Returns the path of the newest schema file, or null if none exists.
+   */
+  private Path newestFile() {
+
+    try {
+      FileStatus[] statuses = rootFileSystem.listStatus(schemaDirectory);
+
+      // No schema files exist, so return null;
+      if (statuses.length == 0) {
+        return null;
+      }
+
+      // Sort the schema files and return the newest one.
+      Arrays.sort(statuses, new FileNameComparator());
+
+      return statuses[statuses.length - 1].getPath();
+
+    } catch (IOException e) {
+      throw new DatasetIOException("Unable to list schema files.", e);
+    }
+  }
+
+  /**
+   * Returns the URI of the newest schema in the manager.
+   */
+  public URI getNewestSchemaURI() {
+
+    Path path = newestFile();
+
+    return path == null ? null : rootFileSystem.makeQualified(path).toUri();
+  }
+
+  private Schema loadSchema(Path schemaPath) {
+    Schema schema = null;
+    InputStream inputStream = null;
+    boolean threw = true;
+
+    try {
+      inputStream = rootFileSystem.open(schemaPath);
+      schema = new Schema.Parser().parse(inputStream);
+      threw = false;
+    } catch (IOException e) {
+      throw new DatasetIOException(
+              "Unable to load schema file:" + schemaPath, e);
+    } finally {
+      try {
+        Closeables.close(inputStream, threw);
+      } catch (IOException e) {
+        throw new DatasetIOException("Cannot close", e);
+      }
+    }
+
+    return schema;
+  }
+
+  /**
+   * Gets the newest schema version being managed.
+   *
+   * @return thew newest schema version
+   */
+  public Schema getNewestSchema() {
+    Path schemaPath = newestFile();
+
+    return schemaPath == null ? null : loadSchema(schemaPath);
+  }
+
+  /**
+   * Imports an existing schema stored at the given path. This
+   * is generally used to bring in schemas written by previous
+   * versions of this library.
+   *
+   * @param schemaPath A path to a schema to import
+   * @return The URI of the schema file managed by this manager.
+   */
+  public URI importSchema(Path schemaPath) {
+
+    Schema schema = loadSchema(schemaPath);
+
+    return writeSchema(schema);
+  }
+
+  /**
+   * Writes the schema and a URI to that schema and returns a URI to the
+   * schema file it writes. The URI can be used in the Hive metastore or
+   * other tools.
+   *
+   * @param schema the schema to write
+   * @return A URI pointing to the written schema
+   */
+  public URI writeSchema(Schema schema) {
+
+    Path previous = newestFile();
+
+    // Ensure all previous schemas are compatible with the new one.
+    Map<Integer, Schema> schemas = getSchemas();
+
+    for (Schema oldSchema: schemas.values()) {
+      if (!SchemaValidationUtil.canRead(oldSchema, schema)) {
```

Comment:
It would be good to have a comment explaining why all previous versions are checked. I think this is because we have to check the case where deleted columns are not compatible with the new schema, right?